### PR TITLE
TrackingCopy moved to storage crate

### PR DESCRIPTION
**Pull request by EdHastingsCasperAssociation**
This PR primarily moves the TrackingCopy from execution engine to storage. Some other affected / supporting types that were in ee have also been moved to storage or to types.

### CODE REVIEW

File: execution_engine/src/engine_state/mod.rs [hunk 1]
```diff
@@ -2192,20 +2214,20 @@ where
     }
 
     /// Gets a trie object for given state root hash.
-    pub fn get_trie_full(&self, trie_key: Digest) -> Result<Option<TrieRaw>, Error>
-    where
-        Error: From<S::Error>,
-    {
-        Ok(self.state.get_trie_full(&trie_key)?)
+    pub fn get_trie_full(&self, trie_key: Digest) -> Result<Option<TrieRaw>, Error> {
+        match self.state.get_trie_full(&trie_key) {
```
Comment (darthsiroftardis) [inline execution_engine/src/engine_state/mod.rs:2218]: Nit: How come we can't use the `?` operator here?

File: execution_engine/src/engine_state/mod.rs [hunk 2]
```diff
@@ -2192,20 +2214,20 @@ where
...
 
     /// Puts a trie if no children are missing from the global state; otherwise reports the missing
     /// children hashes via the `Error` enum.
-    pub fn put_trie_if_all_children_present(&self, trie_bytes: &[u8]) -> Result<Digest, Error>
-    where
-        Error: From<S::Error>,
-    {
-        let missing_children = self.state.missing_children(trie_bytes)?;
+    pub fn put_trie_if_all_children_present(&self, trie_bytes: &[u8]) -> Result<Digest, Error> {
+        let missing_children = match self.state.missing_children(trie_bytes) {
```
Comment (darthsiroftardis) [inline execution_engine/src/engine_state/mod.rs:2227]: Nit: Same here, could we use the `?` operator

File: execution_engine/src/runtime/standard_payment_internal.rs
Comment (darthsiroftardis) [inline execution_engine/src/runtime/standard_payment_internal.rs:1]: We shouldn't need this file soon enough

File: execution_engine_testing/test_support/src/wasm_test_builder.rs [hunk 3]
```diff
@@ -256,12 +251,14 @@ impl LmdbWasmTestBuilder {
                 .expect("should create LmdbTrieStore"),
         );
 
-        let global_state =
-            LmdbGlobalState::empty(environment, trie_store).expect("should create LmdbGlobalState");
+        let max_query_depth = DEFAULT_MAX_QUERY_DEPTH;
```
Comment (darthsiroftardis) [inline execution_engine_testing/test_support/src/wasm_test_builder.rs:254]: Nit: Could we maybe get this value similar to how we derive the `PRODUCTION_GENESIS_REQUEST` to minimize drift between chainspec and default values

File: storage/src/global_state/state/lmdb.rs [hunk 4]
```diff
@@ -1,15 +1,17 @@
 use std::{collections::HashMap, ops::Deref, sync::Arc};
 
 use lmdb::DatabaseFlags;
+
```
Comment (darthsiroftardis) [inline storage/src/global_state/state/lmdb.rs:4]: Nit: No need for this new line?

File: storage/src/global_state/state/lmdb.rs [hunk 5]
```diff
@@ -1,15 +1,17 @@
...
 use tempfile::TempDir;
 
+use crate::tracking_copy::TrackingCopy;
```
Comment (darthsiroftardis) [inline storage/src/global_state/state/lmdb.rs:7]: Nit: I think we can move this import

File: storage/src/global_state/state/mod.rs [hunk 6]
```diff
@@ -70,34 +81,89 @@ pub enum CommitError {
...
+                    Err(err) => BalanceResult::Failure(err),
+                }
+            }
+            Ok(None) => BalanceResult::RootNotFound,
+            Err(err) => BalanceResult::Failure(TrackingCopyError::Storage(err)),
+        }
+    }
+
+    fn transfer(&self, _transfer_request: TransferRequest) -> TransferResult {
+        unimplemented!()
```
Comment (darthsiroftardis) [inline storage/src/global_state/state/mod.rs:161]: Im assuming this where we layer on the additional PRs

File: storage/src/system/standard_payment.rs
Comment (darthsiroftardis) [inline storage/src/system/standard_payment.rs:1]: Hmmm? Doesn't matter since we're getting rid of the contract anyway

**Review by darthsiroftardis — APPROVED** [review]
Some nit picks, but overall its good

File: execution_engine/src/engine_state/mod.rs (hunk 1 above)
Comment (EdHastingsCasperAssociation) [reply execution_engine/src/engine_state/mod.rs:2218]: this function will be deleted in an upcoming PR anyway.

File: execution_engine/src/runtime/standard_payment_internal.rs
Comment (EdHastingsCasperAssociation) [reply execution_engine/src/runtime/standard_payment_internal.rs:1]: yep

File: execution_engine/src/engine_state/mod.rs (hunk 2 above)
Comment (EdHastingsCasperAssociation) [reply execution_engine/src/engine_state/mod.rs:2227]: this function will be deleted in an upcoming PR anyway.

File: execution_engine_testing/test_support/src/wasm_test_builder.rs (hunk 3 above)
Comment (EdHastingsCasperAssociation) [reply execution_engine_testing/test_support/src/wasm_test_builder.rs:254]: maybe...ill look at it in the upcoming PR that insinuates the data provider into the test builder.

File: storage/src/global_state/state/lmdb.rs (hunk 5 above)
Comment (EdHastingsCasperAssociation) [reply storage/src/global_state/state/lmdb.rs:7]: will do on subsequent PR

File: storage/src/system/standard_payment.rs
Comment (EdHastingsCasperAssociation) [reply storage/src/system/standard_payment.rs:1]: yep

**Pull request merged**